lib/binaries/config_source.ts

```diff
@@ -0,0 +1,189 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as request from 'request';
+import * as semver from 'semver';
+import * as xml2js from 'xml2js';
+
+import {Config} from '../config';
+import {HttpUtils} from '../http_utils';
+
+export abstract class ConfigSource {
+  ostype = Config.osType();
+  osarch = Config.osArch();
+  out_dir: string = Config.getSeleniumDir();
+  opt_ignoreSSL: boolean;
+  opt_proxy: string;
+
+  abstract getUrl(version: string): Promise<{url: string, version: string}>;
+  abstract getVersionList(): Promise<string[]>;
+}
+
+export abstract class XmlConfigSource extends ConfigSource {
+  xml: any;
+  xmlUrl: string;
+  fileName: string;
+
+  constructor(private name: string) {
+    super();
+  }
+
+  protected getFileName(): string {
+    try {
+      fs.statSync(this.out_dir);
+    } catch (e) {
+      fs.mkdirSync(this.out_dir);
+    }
+    this.fileName = path.resolve(this.out_dir, this.name + '-response.xml');
+    return this.fileName;
+  }
+
+  protected getXml(): Promise<any> {
+    this.getFileName();
+    let content = this.readResponse();
+    if (content != null) {
+      this.xml = content;
```

Comment:
Again, don't set a property, just return.  Side effects are bad and hard to follow